models/Group.ts

```diff
@@ -126,6 +142,165 @@ export default function (sequelize, DataTypes): GroupStatic {
+   * If there are existing stations that are not in the new set, those stations will be retired.
+   * Any new stations will be added.
+   *
+   * If there is an `applyToRecordingFromDate` Date provided, recordings belonging to this group
+   * will be matched against the new list of stations to see if they fall within the station radius.
+   *
+   * As designed, this will *always* be a bulk import operation of external data from trap.nz
+   *
+   * Returns ids of updated or added stations
+   *
+   */
+  Group.addStationsToGroup = async function (
+    authUser,
+    group,
+    stationsToAdd,
+    applyToRecordingsFromDate
+  ): Promise<StationId[]> {
+    if (!(await group.userPermissions(authUser)).canAddStations) {
+      throw new AuthorizationError(
+        "User doesn't belong to group and/or is not a group admin so cannot add stations"
+      );
+    }
+
+    // Or maybe we should do all of this on the front-end, and show the list of
+    // what will be changed there?
+
+    // Enforce name uniqueness to group here:
+    let existingStations: Station[] = await group.getStations();
+    // Filter out retired stations.
+    existingStations = existingStations.filter(station => station.retiredAt === null);
+
+    const existingStationsByName: Record<string, Station> = {};
+    const newStationsByName: Record<string, CreateStationData> = {};
+    const stationOpsPromises = [];
+    for (const station of stationsToAdd) {
+      newStationsByName[station.name] = station;
+    }
+
+    // Make sure existing stations that are not in the current update are retired, and removed from
+    // the list of existing stations that we are comparing with.
+    const retirePromises = [];
+    const numExisting = existingStations.length;
+    for (let i = 0; i < numExisting; i++) {
+      const station = existingStations.pop();
+      if (!newStationsByName.hasOwnProperty(station.name)) {
+        station.retiredAt = new Date();
+        station.lastUpdatedById = authUser.id;
+        retirePromises.push(station.save());
+      } else {
+        existingStations.unshift(station);
+      }
+    }
+    for (const station of existingStations) {
+      existingStationsByName[station.name] = station;
+    }
+
+    // Make sure no two stations are too close to each other:
+    const allStations = [...existingStations, ...stationsToAdd].map(s => {
+      if (s.hasOwnProperty('lat')) {
+        return {
+          name: (s as CreateStationData).name,
+          location: [(s as CreateStationData).lat, (s as CreateStationData).lng] as [number, number]
+        };
+      } else {
+        return {
+          name: (s as Station).name,
+          location: (s as Station).location.coordinates
+        }
+      }
+    });
+    for (const a of allStations) {
+      for (const b of allStations) {
+        if (a !== b && a.name !== b.name) {
+          if (latLngApproxDistance(a.location, b.location) < MAX_STATION_DISTANCE_THRESHOLD_METERS) {
```

Comment:
would MIN_STATION_SEPARATION be a better name for MAX_STATION_DISTANCE_THRESHOLD_METERS